File: source/com/scaniatv/CustomAPI.java
/*
 * Copyright (C) 2016-2022 phantombot.github.io/PhantomBot
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
// This class is made to be used in the scripts to call APIs. It will make it way more simple with $.customAPI
package com.scaniatv;

import com.gmt2001.HttpRequest;
import com.gmt2001.HttpResponse;
import java.util.HashMap;
import org.json.JSONException;
import org.json.JSONObject;

public class CustomAPI {

    private static CustomAPI instance;

    /*
     * Method to the this instance.
     *
     * @return
     */
    public static synchronized CustomAPI instance() {
        if (instance == null) {
            instance = new CustomAPI();
        }

        return instance;
    }

    /*
     * Class constructor.
     */
    private CustomAPI() {
        Thread.setDefaultUncaughtExceptionHandler(com.gmt2001.UncaughtExceptionHandler.instance());
    }

    /*
     * Method to get a JSON Object from an API.
     *
     * @param  {String} url
     * @return
     */
    public JSONObject getJSON(String url) throws JSONException {
        try {
            HttpResponse data = HttpRequest.getData(HttpRequest.RequestType.GET, url, "", new HashMap<>());
            if (data.success) {
                return new JSONObject(data.content);
            } else {
                throw new JSONException(data.httpCode + ": " + data.exception);
            }
        } catch (JSONException ex) {
            com.gmt2001.Console.err.printStackTrace(ex);
        }
        return new JSONObject("{}");
    }

    /*
     * Method to get data from an API.
     *
     * @param  {String} url
     * @return
     */
    public HttpResponse get(String url) {
        return HttpRequest.getData(HttpRequest.RequestType.GET, url, "", new HashMap<>());
    }

    /*
     * Method to post to an API.
     *
     * @param  {String} url
     * @param  {String} content
     * @return
     */
    public HttpResponse post(String url, String content) {
        return HttpRequest.getData(HttpRequest.RequestType.POST, url, content, new HashMap<>());
    }

    /*
     * Method to put to an API.
     *
     * @param  {String} url
     * @param  {String} content
     * @return
     */
    public HttpResponse put(String url, String content) {
        return HttpRequest.getData(HttpRequest.RequestType.PUT, url, content, new HashMap<>());
    }

    /*
     * Method to delete on an API.
     *
     * @param  {String} url
     * @param  {String} content
     * @return
     */
    public HttpResponse del(String url, String content) {
        return HttpRequest.getData(HttpRequest.RequestType.DELETE, url, content, new HashMap<>());
    }
}

File: javascript-source/core/transformers/customapi.js
/*
 * Copyright (C) 2016-2023 phantombot.github.io/PhantomBot
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* global Packages */

(function () {
    /*
     * @function getCustomAPIValue
     *
     * @param {string} url
     * @returns {string}
     */
    function getCustomAPIValue(url) {
        var res = $.customAPI.get(url);

        if (res.content !== null) {
            $.consoleDebug(res.toString());
            return res.content;
        } else {
            $.log.error(res.toString());
            throw res.toString();
        }
    }

    /*
     * @transformer customapi
     * @formula (customapi url:str) http GET url and output returned text
     * @labels twitch discord commandevent customapi
     * @notes the command tag (token) can be placed in the url for a secret token saved via !tokencom or the panel
     * @notes if any args, $1-$9, are used in the url, they are required to be provided by the user issuing the command or the tag will abort and return an error message instead
     * @notes this will output the full response from the remote url, so be careful not to cause spam or lock up the bot with a webpage
     * @example Caster: !addcom !joke (customapi http://not.real.com/joke.php?name=$1)
     * User: !joke bear
     * Bot: These jokes are un-bear-able
     */
    function customapi(args) {
        var match;
        if ((match = args.args.match(/^\s(.+)$/))) {
            var cmd = args.event.getCommand();
            var flag = false;
            match[1] = match[1].replace(/\$([1-9])/g, function (m) {
                var i = parseInt(m[1]);
                if (!args.event.getArgs()[i - 1]) {
                    flag = true;
                    return m[0];
                }
                return args.event.getArgs()[i - 1];
            });
            if (flag) {
                return {result: $.lang.get('customcommands.customapi.404', cmd)};
            }
            var response;
            try {
                response = getCustomAPIValue(match[1]);
            } catch (ex) {
                return {result: $.lang.get('customcommands.customapijson.err', cmd)};
            }
            return {
                result: response,
                cache: false
            };
        }
    }

    /*
     * @transformer customapijson
     * @formula (customapijson url:str specs:str) httpGet url and extract json info according to specs
     * @labels twitch discord commandevent customapi
     * @notes the command tag (token) can be placed in the url for a secret token saved via !tokencom or the panel
     * @notes if any args, $1-$9, are used in the url, they are required to be provided by the user issuing the command or the tag will abort and return an error message instead
     * @notes the response must be a JSONObject. arrays are only supported with a known index, walking arrays is not supported
     * @notes multiple specs can be provided, separated by spaces; curly braces can be used to enclose literal strings
     * @example Caster: !addcom !weather (customapijson http://api.apixu.com/v1/current.json?key=NOT_PROVIDED&q=$1 {Weather for} location.name {:} current.condition.text {Temps:} current.temp_f {F} current.temp_c {C})
     * User: !weather 80314
     * Bot: Weather for Boulder, CO : Sunny Temps: 75 F 24 C
     */
    var reCustomAPITextTag = new RegExp(/{([\w\W]+)}/);
    var JSONObject = Packages.org.json.JSONObject;
    function customapijson(args, event) {
        var match,
                customJSONStringTag,
                jsonCheckList,
                jsonItems,
                jsonObject,
                response,
                responsePart;
        if ((match = args.args.match(/^ (\S+) (.+)$/))) {
            var cmd = args.event.getCommand();
            var flag = false;
            match[1] = match[1].replace(/\$([1-9])/g, function (m) {
                var i = parseInt(m[1]);
                if (!args.event.getArgs()[i - 1]) {
                    flag = true;
                    return m[0];
                }
                return args.event.getArgs()[i - 1];
            });
            if (flag) {
                return {result: $.lang.get('customcommands.customapi.404', cmd)};
            }

            var result = '';
            try {
                response = getCustomAPIValue(match[1]);
            } catch (ex) {
                return {result: $.lang.get('customcommands.customapijson.err', cmd)};
            }
            jsonItems = match[2].split(' ');
            for (var j = 0; j < jsonItems.length; j++) {
                if (jsonItems[j].startsWith('{') && jsonItems[j].endsWith('}')) {
                    result += " " + jsonItems[j].match(reCustomAPITextTag)[1];
                } else if (jsonItems[j].startsWith('{') && !jsonItems[j].endsWith('}')) {
                    customJSONStringTag = '';
                    while (!jsonItems[j].endsWith('}')) {
                        customJSONStringTag += jsonItems[j++] + " ";
                    }
                    customJSONStringTag += jsonItems[j];
                    result += " " + customJSONStringTag.match(reCustomAPITextTag)[1];
                } else {
                    jsonCheckList = jsonItems[j].split('.');
                    if (jsonCheckList.length === 1) {
                        try {
                            responsePart = new JSONObject(response).get(jsonCheckList[0]);
                        } catch (ex) {
                            $.log.error('Failed to get data from API: ' + ex);
                            $.log.error(response.toString());
                            return {result: $.lang.get('customcommands.customapijson.err', cmd)};
                        }
                        result += responsePart;
                    } else {
                        for (i = 0; i < jsonCheckList.length - 1; i++) {
                            if (i === 0) {
                                try {
                                    jsonObject = new JSONObject(response).get(jsonCheckList[i]);
                                } catch (ex) {
                                    $.log.error('Failed to get data from API: ' + ex);
                                    $.log.error('response:' + response.toString());
                                    return {result: $.lang.get('customcommands.customapijson.err', cmd)};
                                }
                            } else if (!isNaN(jsonCheckList[i + 1])) {
                                try {
                                    jsonObject = jsonObject.get(jsonCheckList[i]);
                                } catch (ex) {
                                    $.log.error('Failed to get data from API: ' + ex);
                                    $.log.error('jsonCheckList[' + i + ']: ' + response.toString());
                                    return {result: $.lang.get('customcommands.customapijson.err', cmd)};
                                }
                            } else {
                                try {
                                    jsonObject = jsonObject.get(jsonCheckList[i]);
                                } catch (ex) {
                                    $.log.error('Failed to get data from API: ' + ex);
                                    $.log.error('jsonCheckList[' + i + ']: ' + jsonCheckList[i]);
                                    return {result: $.lang.get('customcommands.customapijson.err', cmd)};
                                }
                            }
                        }
                        try {
                            responsePart = jsonObject.get(jsonCheckList[i]);
                        } catch (ex) {
                            $.log.error('Failed to get data from API: ' + ex);
                            $.log.error('jsonCheckList[' + i + ']: ' + jsonCheckList[i]);
                            return {result: $.lang.get('customcommands.customapijson.err', cmd)};
                        }
                        result += responsePart;
                    }
                }
            }

            return {
                result: result,
                cache: false
            };
        }
    }

    var transformers = [
        new $.transformers.transformer('customapi', ['twitch', 'discord', 'commandevent', 'customapi'], customapi),
        new $.transformers.transformer('customapijson', ['twitch', 'discord', 'commandevent', 'customapi'], customapijson)
    ];

    $.transformers.addTransformers(transformers);
})();
